aiida/backends/tests/tools/graph/test_age.py

```diff
@@ -0,0 +1,771 @@
+        expect0 = set([data_node.id, work_node.id])
+        self.assertEqual(result0, expect0)
+
+        input_nodes = [data_node.id]
+        basket0 = Basket(nodes=input_nodes)
+        queryb0 = QueryBuilder().append(Node).append(Node)
+
+        rerule1 = ReplaceRule(queryb0, max_iterations=1)
+        result1 = rerule1.run(basket0.copy())['nodes'].keyset
+        self.assertEqual(result1, set([work_node.id]))
+
+        rerule2 = ReplaceRule(queryb0, max_iterations=2)
+        result2 = rerule2.run(basket0.copy())['nodes'].keyset
+        self.assertEqual(result2, set([data_node.id]))
+
+        rerule3 = ReplaceRule(queryb0, max_iterations=3)
+        result3 = rerule3.run(basket0.copy())['nodes'].keyset
+        self.assertEqual(result3, set([work_node.id]))
+
+        rerule4 = ReplaceRule(queryb0, max_iterations=4)
+        result4 = rerule4.run(basket0.copy())['nodes'].keyset
+        self.assertEqual(result4, set([data_node.id]))
+
+    @staticmethod
+    def _create_branchy_graph():
+        """
+        Creates a basic branchy graph which has two concatenated calculations:
+
+        * CN1 takes DN0 as inputs and returns DN1 and DNO.
+        * CN2 takes DN1 and DNI as inputs and returns DN2.
+
+        This graph looks like this::
+
+                        +-----+                 +-----+
+                        | DNO |                 | DNI |
+                        +-----+                 +-----+
+                           ^                       |
+                           |                       v
+            +-----+     +-----+     +-----+     +-----+     +-----+
+            | DN0 | --> | CN1 | --> | DN1 | --> | CN2 | --> | DN2 |
+            +-----+     +-----+     +-----+     +-----+     +-----+
+        """
+        from aiida import orm
+
+        dn0 = orm.Data().store()
+        cn1 = orm.CalculationNode()
+        cn1.add_incoming(dn0, link_type=LinkType.INPUT_CALC, link_label='inpcalc_dn0')
+        cn1.store()
+
+        dn1 = orm.Data()
+        dno = orm.Data()
+        dn1.add_incoming(cn1, link_type=LinkType.CREATE, link_label='create_dn1')
+        dno.add_incoming(cn1, link_type=LinkType.CREATE, link_label='create_dno')
+        dn1.store()
+        dno.store()
+
+        dni = orm.Data().store()
+        cn2 = orm.CalculationNode()
+        cn2.add_incoming(dn1, link_type=LinkType.INPUT_CALC, link_label='inpcalc_dn1')
+        cn2.add_incoming(dni, link_type=LinkType.INPUT_CALC, link_label='inpcalc_dni')
+        cn2.store()
+
+        dn2 = orm.Data()
+        dn2.add_incoming(cn2, link_type=LinkType.CREATE, link_label='create_dn2')
+        dn2.store()
+
+        output_dict = {}
+        output_dict['DNI'] = dni
+        output_dict['DN0'] = dn0
+        output_dict['DN1'] = dn1
+        output_dict['DN2'] = dn2
+        output_dict['DNO'] = dno
+        output_dict['CN1'] = cn1
+        output_dict['CN2'] = cn2
+        return output_dict
```

Comment:
I changed the way I defined the `output_dict`, but I'm a bit hesitant to put that directly in the `return`. I mean, it would be ok if I could define the whole dict in one short line, but having a multiline `return` feels so wrong. Is there a reason why it would be better that way? Surely the temporary variable has to be created anyways.

Reply:
> (...) but having a multiline `return` feels so wrong. Is there a reason why it would be better that way? Surely the temporary variable has to be created anyways.

The temporary variable does not have to be created, but it will probably take the same amount of memory.
Multiline returns are not bad, far from it, in my view.
To me it just feels more wasteful to create a variable and then immediately return it without using it - then I'd rather just return the value in the first place. Skip the intermediate step, so to speak.